enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,535 @@
+
+### Why allow operators to communicate status through CustomResources
+
+There are many reasons why operators should communicate their conditions through owned CustomResources:
+
+- CustomResources are the main entry point to an operator and operators are already expected to update the status of the CustomResource to communicate "Actual State".
+- By writing the condition to the status of an owned CustomResource, it becomes very clear which CR is responsible for changing OLM behavior.
+- Introducing an OLM owned CRD forces operators to integrate with OLM at a much deeper level than writing to a resource that they already manage. Operators would be required to detect if it is being managed by OLM and begin writing to the API. By writing to an owned CR, the Operator does not require OLM specific code and can be run independently of OLM.
+
+Allowing operators to write to the status of any CR creates a potential scaling issue if enough CRs are on cluster. OLM will avoid these performance issues by introducing the Status Probe Operator (SPO) which will reconcile the `probe` resource. If an operator chooses to write to their own CRs, SPO will create a pod that will keep the status of the `probe` resource synced with the conditions in the CRs.
+
+### Example
+
+Operators may opt-in to "OLM Supported Conditions" by adding an annotation to their owned CRD, where the key is the defined by OLM and the value is the `type` of the condition on the CR.
+
+The "OLM Supported Condition" annotation key will adhere to the following format: `olm.Condition.<ConditionName>`. The value of the annotation will match the `Type` of the condition. For example, the `Upgradeable` condition's annotation key will be `olm.Condition.Upgradeable`.
+
+"OLM Supported Condition" annotation keys will map to multiple condition types separated by commas if needed.
+
+Let's consider an operator that installs a database. Assume that the operator already communicates that a database migration is in progress by setting the "Migrating" condition type in the status of the `foo` CustomResourceDefinition and the operator does not want to upgrade when a database is Migrating. In this case, the operator's CRD should include the following annotations:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata: {
+  name: foo
+  annotations: {
+    "olm.Condition.Upgradeable" : "Migrating"
+    "olm.Condition.Important" : "BadConnectivity,UnhealthyDatabase"
+  }
+}
+```
+
+When this annotation is present on a CRD, OLM will append the following `Probe` CR to the `installPlan`:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # Changes to the CRD's annotations should override the spec. SPO will watch the CRD for changes to the "OLM Supported Condition" annotations and update the spec accordingly.
+  probeResources:
+  - name: foo.example.com 
```

Comment:
Is this only updated from a CRD's annotations a matching <kind>.<group> is added here?

Reply:
Because of the `spec.specOverride: crdAnnotations` field above, the array of `probeResources` will be managed by the SPO.

Reply:
let me rephrase: new CRDs cannot be discovered dynamically without updating this spec.

Is this an accurate statement?

Reply:
I have not planned to implement logic such that new CRDs are automatically discovered. OLM has typically taken the approach that the operator does not introduce CRDs itself.